src/Integration.Vsix/SonarLintTracker/IssueTracker.cs

```diff
@@ -0,0 +1,182 @@
+    ///</summary>
+    /// <remarks><para>The lifespan of this object is tied to the lifespan of the taggers on the view. On creation of the first tagger,
+    /// it starts tracking errors. On the disposal of the last tagger, it shuts down.</para>
+    /// </remarks>
+    public class IssueTracker : ITagger<IErrorTag>, IDisposable
+    {
+        private readonly TaggerProvider provider;
+        private readonly ITextBuffer textBuffer;
+
+        private ITextSnapshot currentSnapshot;
+        private NormalizedSnapshotSpanCollection dirtySpans;
+
+        internal readonly string FilePath;
+        internal readonly SnapshotFactory Factory;
+
+        internal IssuesSnapshot snapshot;
+
+        internal IssueTracker(TaggerProvider provider, ITextBuffer buffer, string filePath)
+        {
+            this.provider = provider;
+            this.textBuffer = buffer;
+            this.currentSnapshot = buffer.CurrentSnapshot;
+
+            this.FilePath = filePath;
+            this.Factory = new SnapshotFactory(new IssuesSnapshot(this.FilePath, 0));
+
+            this.Init();
+        }
+
+        internal void Init()
+        {
+            textBuffer.ChangedLowPriority += this.OnBufferChange;
+            provider.AddIssueTracker(this);
+            dirtySpans = new NormalizedSnapshotSpanCollection();
+
+            SonarLintDaemonIntegrationPackage.Instance.RequestAnalysis(FilePath);
+        }
+
+        public void Dispose()
+        {
+            textBuffer.ChangedLowPriority -= this.OnBufferChange;
+            provider.RemoveIssueTracker(this);
+        }
+
+        private void OnBufferChange(object sender, TextContentChangedEventArgs e)
+        {
+            UpdateDirtySpans(e);
+
+            var newMarkers = TranslateMarkerSpans();
+
+            SnapToNewSnapshot(newMarkers);
+        }
+
+        private void UpdateDirtySpans(TextContentChangedEventArgs e)
+        {
+            currentSnapshot = e.After;
+
+            // Translate all of the old dirty spans to the new snapshot.
+            NormalizedSnapshotSpanCollection newDirtySpans = dirtySpans.CloneAndTrackTo(e.After, SpanTrackingMode.EdgeInclusive);
+
+            // Dirty all the spans that changed.
+            foreach (var change in e.Changes)
+            {
+                newDirtySpans = NormalizedSnapshotSpanCollection.Union(newDirtySpans, new NormalizedSnapshotSpanCollection(e.After, change.NewSpan));
+            }
+
+            dirtySpans = newDirtySpans;
+        }
+
+        // Translate spans to the updated snapshot of the same ITextBuffer
+        private IssuesSnapshot TranslateMarkerSpans()
+        {
+            var oldMarkers = this.Factory.CurrentSnapshot;
+            var newMarkers = new IssuesSnapshot(this.FilePath, oldMarkers.VersionNumber + 1);
+
+            // Copy all of the old errors to the new errors unless the error was affected by the text change
+            foreach (var marker in oldMarkers.IssueMarkers)
```

Comment:
this could become:

    newMarkers.IssueMarkers.AddRange(
        oldMarkers.IssueMarkers
            .Select(marker => IssueMarker.CloneAndTranslateTo(marker, currentSnapshot))
            .Where(clone => clone != null));